adapters/lora/src/main/java/org/eclipse/hono/adapter/lora/LoraCommandSubscriptions.java

```diff
@@ -0,0 +1,179 @@
+        this(vertx, null);
+    }
+
+    /**
+     * Creates a new Lora command subscriptions class.
+     *
+     * @param vertx The Vert.x instance to use.
+     * @param tracer The tracer instance.
+     */
+    public LoraCommandSubscriptions(final Vertx vertx, final Tracer tracer) {
+        this.commandSubscriptions = new ConcurrentHashMap<>();
+        this.vertx = vertx;
+        this.tracer = tracer == null ? NoopTracerFactory.create() : tracer;
+    }
+
+    /**
+     * Subscribes for tenant diasbled/deleted events, in order to close the consumer objects associated with them later.
+     */
+    public void init() {
+        NotificationEventBusSupport.registerConsumer(vertx, TenantChangeNotification.TYPE,
+                notification -> {
+                    if (LifecycleChange.DELETE.equals(notification.getChange())
+                            || (LifecycleChange.UPDATE.equals(notification.getChange())
+                                    && !notification.isTenantEnabled())) {
+                        closeConsumersForTenant(notification.getTenantId());
+                    }
+                });
+
+        NotificationEventBusSupport.registerConsumer(vertx, DeviceChangeNotification.TYPE,
+                notification -> {
+                    if (LifecycleChange.DELETE.equals(notification.getChange())) {
+                        closeConsumersForGateway(notification.getTenantId(), notification.getDeviceId());
+                    }
+                });
+    }
+
+    private void closeConsumersForGateway(final String tenantId, final String gatewayId) {
+        final SubscriptionKey key = new SubscriptionKey(tenantId, gatewayId);
+        final TriTuple<CommandConsumer, LoraProvider, Context> subscription = commandSubscriptions.remove(key);
+        if (subscription == null) {
+            return;
+        }
+
+        final Context currentCtx = vertx.getOrCreateContext();
+        final Span span = TracingHelper
+                .buildSpan(tracer, null, "close command subscription for deleted gateway",
+                        getClass().getSimpleName())
+                .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_CLIENT)
+                .start();
+        TracingHelper.setDeviceTags(span, tenantId, gatewayId);
+        final Context subscriptionContext = subscription.three();
+        subscriptionContext.runOnContext(x -> {
+            final CommandConsumer commandConsumer = subscription.one();
+            commandConsumer.close(span.context()).onComplete(r -> {
+                currentCtx.runOnContext(s -> span.finish());
+            });
+        });
+    }
+
+    @SuppressWarnings("rawtypes")
+    private void closeConsumersForTenant(final String tenantId) {
+        if (!commandSubscriptions.entrySet().stream().anyMatch(e -> e.getKey().getTenant().equals(tenantId))) {
+            return;
+        }
+
+        final Context currentCtx = vertx.getOrCreateContext();
+        final Span span = TracingHelper
```

Comment:
I was trying to test this by running the `LoraIT` (manuall making sure that a device with command endpoint is created, so that command consumer creation isn't skipped), but I didn't see this span in the Jaeger UI when the tenant got deleted after the test. @kduba Could you check this?

Reply:
I've debugged this. The reason for the missing span is that the code execution never gets into this new class. In the _registerCommandConsumerIfNeeded()_ in _LoraAdapterProtocol_ class, _CommandEndpoint_ seems to be null after the assertion of the registration, and falls in:
```
.compose(assertion -> {
                    if (assertion.getCommandEndpoint() == null) {
                        LOG.debug("gateway has no command endpoint defined, skipping command consumer creation [tenant: {}, gateway-id: {}]",
                                tenantId, gatewayId);
                        currentSpan.log("gateway has no command endpoint defined, skipping command consumer creation");
                        return Future.succeededFuture((Void) null);
                    }
```
That means that no commandConsumer is created, and this new class is not used at all.. The same happens with all _LoraIT_ tests.


There are 2 classes implementing _DeviceRegistrationClient_ interface - _ProtonBasedDeviceRegistrationClient_ and _JmsBasedRegistrationClient_. I assume _JmsBasedRegistrationClient_ is used for the tests and the other for real use, maybe that's where the difference in behavior comes from? But anyway, I don't see how I can change this commandEndpoint to not be null from the integration tests level.. or can I?

Reply:
> _CommandEndpoint_ seems to be null after the assertion of the registration

You could check where for the `LoraIT` tests the `org.eclipse.hono.service.management.device.Device` instances get created (see `helper.registry.addDeviceForTenant()`) and set a `CommandEndpoint` instance on it there for local testing. (You could check other places in the code base where `device.setCommandEndpoint(commandEndpoint)` gets invoked for `CommandEndpoint` examples.)

Reply:
I added a `commandEndpoint` but it didn't change the behavior, so I just skipped the null check so I could test the code..

- The `ConcurrentHashMap<>` doesn't seem to support the remove operation in the iterator, so I returned the removing from the map as a separate step.

```
ERROR [io.qua.ver.cor.run.VertxCoreRecorder] Uncaught exception received by Vert.x: java.lang.UnsupportedOperationException: remove
at java.base/java.util.Iterator.remove(Unknown Source)
```

- I checked both contexts (on span create and finish) and they are the same. So the context is not the reason for the span not appearing. Maybe it's because we are using Noop tracer? I also changed the tracer to be set to the one from the LoraAdaper, I think it makes more sense for them to be the same.

- This exception was thrown on close consumer (with noop span context), so I added a boolean variable, to know when to pass null as span context:
```
[io.qua.ver.cor.run.VertxCoreRecorder] (vert.x-eventloop-thread-2) Uncaught exception received by Vert.x: 
java.lang.IllegalArgumentException: context is not a valid SpanContextShim object: io.opentracing.noop.NoopSpanContextImpl
```

Reply:
> I added a `commandEndpoint` but it didn't change the behavior,

For a local test, using something like
```
        final Device device = new Device();
        final CommandEndpoint commandEndpoint = new CommandEndpoint();
        commandEndpoint.setHeaders(Map.of("my-header", "my-header-value"));
        commandEndpoint.setUri("https://localhost:11111/commands/{{deviceId}}/send");
        device.setCommandEndpoint(commandEndpoint);
        return addDeviceForTenant(tenantId, tenant, deviceId, device, password);
```
[here](https://github.com/eclipse-hono/hono/blob/a6ce21f8bdf450819b7d122dde1993bf7a2141ff/tests/src/test/java/org/eclipse/hono/tests/DeviceRegistryHttpClient.java#L1083) worked for me.

> Maybe it's because we are using Noop tracer? I also changed the tracer to be set to the one from the LoraAdaper, I think it makes more sense for them to be the same.

Using a Noop tracer of course means that no traces get created. And yes, there should only be _one_ Tracer instance in the application. For the Hono applications, it's the singleton `Tracer` bean created by the `org.eclipse.hono.service.tracing.TracerProducer`. So you have to make sure, that that tracer instance is passed on correctly. Looking at the places where a `LoraCommandSubscriptions` instance gets created, no tracer seems to be passed on, though. So that seems to be the issue here.

> This exception was thrown on close consumer (with noop span context), so I added a boolean variable, to know when to pass null as span context:

This exception is thrown, when you mix using an OpenTelemetry TracerShim Tracer (so, an "actual" Tracer), with another Tracer (here the NoopTracer). That shouldn't be done (see above).
The extra boolean variable isn't needed. If the parent span context was already created by a Noop Tracer, there would be no such exception.

Reply:
> * The `ConcurrentHashMap<>` doesn't seem to support the remove operation in the iterator, so I returned the removing from the map as a separate step.

It's rather the fact that the iterator here is based on a filtered stream of the map entries. If you use `commandSubscriptions.entrySet().iterator()` directly and use the filter condition (`subscription.getKey().getTenant().equals(tenantId)`) in an `if` block inside the loop instead, there is no such exception.

Reply:
```
public void setCommandSubscriptions(final LoraCommandSubscriptions commandSubscriptions) {
        this.commandSubscriptions = commandSubscriptions;
        this.commandSubscriptions.setTracer(tracer);
    }
```
I'm using the setter of the commandSubscriptions. I checked that the tracer is set to the LoraAdapterProtocol, before the commandSubscriptions are set. But if this doesn't look good I'll find another way to set the tracer? That however doesn't seems to resolve the exception, but I'll double check again..

Reply:
Ok, now after the last changes, the iterator is ok, the tracer exception doesn't seem to appear, and the contexts of the span are the same.

Reply:
Concerning the tracer: Doing a local test, I still got the
`Uncaught exception received by Vert.x: java.lang.IllegalArgumentException: context is not a valid SpanContextShim object`
exception in the log.
I think your previous solution, having a _final_ `LoraCommandSubscriptions.tracer` field and setting it in the constructor, was better. You could just make sure that, when invoking the constructor in the `LoraCommandSubscriptionsFactory` class, you pass on the `Tracer` bean instance. That instance can simply be obtained by adding a `tracer` param to the `LoraCommandSubscriptionsFactory.commandSubscriptions` method.

---
Running a local test with that change (tested using `mvn verify -Prun-tests,jaeger,lora-only -Dit.test=LoraIT -Ddocker.keepRunning` and accessing `http://localhost:18080`), I still did not see the "close command subscriptions [...]" span. The reason for that is a bit tricky.
Since the `closeConsumersForTenant`/`closeConsumersForGateway` methods are triggered by a Vert.x event bus message, there will be corresponding tracing spans created for sending/receiving the message (by Vert.x itself). The "close command subscriptions [...]" span will get such a span as parent implicitly.
The Vert.x event bus message spans themselves represent more noise than benefit in Hono, therefore we have globally suppressed the sampling (i.e. storing) of such traces.
To see such spans in local tests, you could disable the code for suppressing the spans by using an empty `ArrayList` [here](https://github.com/eclipse-hono/hono/blob/8ca3b536a9bc7701dd33af91589405d81e1803db/service-base/src/main/java/org/eclipse/hono/service/tracing/SamplerProducer.java#L217).
If you do that for the LoraIT tests, you would see:
![grafik](https://user-images.githubusercontent.com/891931/197452182-777b2ba9-155f-4b1a-9e17-a9878005c436.png)
The `notification.tenant-change-v1 send/receive` spans are the event bus message spans here (with the event bus address). Suppressing these spans means that also the sampling of the child spans get suppressed.
That's why you see no "close command subscriptions [...]" spans.

Solution:
The "close command subscriptions [...]" span gets the event bus message span as parent implicitly because the event bus message span is set by Vert.x as the *active* span of the Vert.x context.
Therefore, clearing the active span will decouple the "close command subscriptions [...]" span.
This can be done with
```
        // Clear currently active span, i.e. the event bus message span, so that it isn't used as parent span.
        // Otherwise the span created here wouldn't get sampled because event bus message spans don't get sampled in Hono.
        tracer.activateSpan(null);
```
before creating the span in `closeConsumersForGateway` and `closeConsumersForTenant`.